File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < t; i++) {
			Stack<String> stack = new Stack<>();
			String[] str = br.readLine().split("");
			boolean check = false;
			for (String s : str) {
				if (s.equals("(")) {
					stack.push(s);
				} else {
					if (stack.isEmpty()) {
						check = true;
						break;
					}
					stack.pop();
				}
			}
			
			if (!stack.isEmpty() || check) {
				sb.append("NO").append("\n");
			} else {
				sb.append("YES").append("\n");
			}
		}
		System.out.println(sb.toString());
	}

}

File: 백준/Silver/9012. 괄호/괄호.py
from sys import stdin

input = stdin.readline


def check(s):
    stack = []
    for i in range(len(s)):
        if s[i] == "(":
            stack.append(s[i])
        else:
            if not stack:
                return "NO"
            stack.pop()
    return "NO" if stack else "YES"


def solve():
    T = int(input())
    for _ in range(T):
        s = input().strip()
        print(check(s))


if __name__ == "__main__":
    solve()
